### PR TITLE
feat: MagicOnion.Server.OpenTelemetry to use OpenTelemetry-dotnet 1.0.0-rc1

### PR DESCRIPTION
**Pull request by guitarrapc**
## TL;DR

* Update package for MagicOnion.Server.OpenTelemetry to use latest OpenTelemetry-dotnet (1.0.0-rc1)
* Update ChatApp.Server.Telemetry sample to select jaeger/zipkin/console for tracer.
* Update ChatApp.Server.Telemetry docker and README.
* Update ChatApp.Unity to use v4.0.1 unity package.
* (BreakingChange) AddMagicOnionOpenTelemetry is changed to AddOpenTelemetry and use IMagicOnionServerBuilder instead of IServiceCollection.
* Remove DynamicCodeDumper.
* Fix NullReferenceException happen when Activity is null. (MagicOnion Tracer Attributes will not run when activity listener or sample is not listen to any.)
* Fix MagicOnion.Experimental.sln failed to build sln due to missing reference to MagicOnion.Server and MagicOnion.Abstractions.

### CODE REVIEW

File: samples/ChatApp.Telemetry/ChatApp.Server/ChatApp.Server.csproj [hunk 1]
```diff
@@ -2,14 +2,16 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
-    <LangVersion>7.3</LangVersion>
+    <TargetFramework>net5.0</TargetFramework>
+    <LangVersion>8.0</LangVersion>
```
Comment (mayuki) [inline samples/ChatApp.Telemetry/ChatApp.Server/ChatApp.Server.csproj:6]: **mosikasite**: `9.0` or `latest`

`LangVersion` is no longer needed in this project.

**Review by mayuki — COMMENTED** [review]
`AddMagicOnionOpenTelemetry` method is an extension to `IServiceCollection` interface. Since v4 it should be an extension to `IMagicOnionServerBuilder`.

- https://github.com/Cysharp/MagicOnion/blob/master/src/MagicOnion.Server.Authentication/Jwt/MagicOnionAuthenticationJwtServiceCollectionExtensions.cs#L20
- https://github.com/Cysharp/MagicOnion/blob/master/src/MagicOnion.Server.Redis/MagicOnionServerBuilderRedisExtensions.cs#L14

**Comment by guitarrapc** [comment]
Is DynamicCodeDumper can convert to net5.0 or netcoreapp3.1?
```
Error	NU1201	Project MagicOnion is not compatible with net471 (.NETFramework,Version=v4.7.1). Project MagicOnion supports:
  - net5.0 (.NETCoreApp,Version=v5.0)
  - netcoreapp3.1 (.NETCoreApp,Version=v3.1)	DynamicCodeDumper	
```

**Comment by mayuki** [comment]
DynamicCodeDumper project is no longer in use and should be removed from the solution.

File: samples/ChatApp.Telemetry/ChatApp.Server/ChatApp.Server.csproj (hunk 1 above)
Comment (guitarrapc) [reply samples/ChatApp.Telemetry/ChatApp.Server/ChatApp.Server.csproj:6]: If I set `<Project Sdk="Microsoft.NET.Sdk.Web">` even TargetGramework is net5.0 but langversion detected as C# 7.3, this behavior is not detected with `Microsoft.NET.Sdk` though.

Comment (guitarrapc) [reply samples/ChatApp.Telemetry/ChatApp.Server/ChatApp.Server.csproj]: set to use latest.

**Comment by guitarrapc** [comment]
Changed to use IMagicOnionServerBuilder

**Review by mayuki — APPROVED** [review]

File: src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs [hunk 2]
```diff
@@ -38,7 +38,6 @@ public override async ValueTask Invoke(ServiceContext context, Func<ServiceConte
         {
             // https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md#grpc
 
-            // Client -> Server incoming filter
             // span name must be `$package.$service/$method` but MagicOnion has no $package.
             using var activity = source.StartActivity($"{context.MethodType}:{context.CallContext.Method}", ActivityKind.Server);
```
Comment (itn3000) [inline src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs:42]: Activity from `ActivitySource.StartActivity` may be null if no one listen or all listener returns ActivitySamplingResult.None.
so it should be considered `activity` may be null.

File: src/MagicOnion.Server.OpenTelemetry/OpenTelemetryServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -55,35 +55,42 @@ public static class OpenTelemetryServiceCollectionExtensions
...
                 });
-                services.AddSingleton(tracerFactory);
-                services.AddSingleton(new ActivitySource(options.ActivitySourceName));
+                builder.Services.AddSingleton(tracerFactory);
+
+                // Avoid directly register ActivitySource to Singleton for easier identification.
+                var activitySource = new ActivitySource(options.MagicOnionActivityName);
+                var magicOnionActivitySources = new MagicOnionActivitySources(activitySource);
+                builder.Services.AddSingleton(magicOnionActivitySources);
             }
```
Comment (guitarrapc) [inline src/MagicOnion.Server.OpenTelemetry/OpenTelemetryServiceCollectionExtensions.cs:85]: Should be better listen by itself when startup.

```csharp
                var activityListener = new ActivityListener
                {
                    ShouldListenTo = s => true,
                    SampleUsingParentId = (ref ActivityCreationOptions<string> activityOptions) => ActivitySamplingResult.AllData,
                    Sample = (ref ActivityCreationOptions<ActivityContext> activityOptions) => ActivitySamplingResult.AllData
                };
                ActivitySource.AddActivityListener(activityListener);
```

File: src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs (hunk 2 above)
Comment (guitarrapc) [reply src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs:42]: thank you, added null check.

Comment (itn3000) [reply src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs:42]: with this fix, Invoke seems to exit without invoke next when Activity is null

**Comment by guitarrapc** [comment]
Now opentelemetry-dotnet 1.0.0-rc1 ready.

File: src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs (hunk 2 above)
Comment (guitarrapc) [reply src/MagicOnion.Server.OpenTelemetry/OpenTelemetryCollectorTracerFilterAttributes.cs:42]: oops, fixed.

**Review by neuecc — APPROVED** [review]
looks good, thanks.
